src/System.Private.CoreLib/shared/System/Runtime/InteropServices/SafeBuffer.cs

```diff
@@ -110,13 +110,20 @@ public void Initialize(ulong numBytes)
         [CLSCompliant(false)]
         public void Initialize(uint numElements, uint sizeOfEachElement)
         {
-            if (IntPtr.Size == 4 && numElements * sizeOfEachElement > uint.MaxValue)
-                throw new ArgumentOutOfRangeException("numBytes", SR.ArgumentOutOfRange_AddressSpace);
+            if (sizeOfEachElement == 0)
+            {
+                _numBytes = (UIntPtr)0;
+            }
+            else
+            {
+                if (IntPtr.Size == 4 && numElements > uint.MaxValue / sizeOfEachElement)
+                    throw new ArgumentOutOfRangeException("numBytes", SR.ArgumentOutOfRange_AddressSpace);
 
-            if (numElements * sizeOfEachElement >= (ulong)Uninitialized)
-                throw new ArgumentOutOfRangeException(nameof(numElements), SR.ArgumentOutOfRange_UIntPtrMax);
+                if (numElements >= (ulong)Uninitialized / sizeOfEachElement)
```

Comment:
This cannot be ever true.

For 64-bit, `uint * uint` is always smaller than `ulong.MaxValue`. 
For 32-bit, it is identical to the condition above.